autogpt_platform/backend/backend/data/dynamic_fields.py

```diff
@@ -137,30 +149,64 @@ def _tokenise(path: str) -> list[tuple[str, str]] | None:
     return tokens
 
 
-def parse_execution_output(output: tuple[str, Any], name: str) -> Any:
+def parse_execution_output(
+    output_item: tuple[str, Any],
+    link_output_selector: str,
+    sink_node_id: str | None = None,
+    sink_pin_name: str | None = None,
+) -> Any:
     """
-    Retrieve a nested value out of `output` using the flattened *name*.
+    Retrieve a nested value out of `output` using the flattened `link_output_selector`.
+
+    On any failure (wrong name, wrong type, out-of-range, bad path) returns **None**.
 
-    On any failure (wrong name, wrong type, out-of-range, bad path)
-    returns **None**.
+    ### Special Case: Tool pins
+    For regular output pins, the `output_item`'s name will simply be the field name, and
+    `link_output_selector` (= the `source_name` of the link) may provide a "selector"
+    used to extract part of the output value and route it through the link
+    to the next node.
+
+    However, for tool pins, it is the other way around: the `output_item`'s name
+    provides the routing information (`tools_^_{sink_node_id}_~_{field_name}`),
+    and the `link_output_selector` is simply `"tools"`
+    (or `"tools_^_{tool_name}_~_{field_name}"` for backward compatibility).
 
     Args:
-        output: Tuple of (base_name, data) representing a block output entry
-        name: The flattened field name to extract from the output data
+        output_item: Tuple of (base_name, data) representing a block output entry.
+        link_output_selector: The flattened field name to extract from the output data.
+        sink_node_id: Sink node ID, used for tool use routing.
+        sink_pin_name: Sink pin name, used for tool use routing.
 
     Returns:
-        The value at the specified path, or None if not found/invalid
+        The value at the specified path, or `None` if not found/invalid.
     """
-    base_name, data = output
+    output_pin_name, data = output_item
+
+    # Special handling for tool pins
+    if is_tool_pin(link_output_selector) and (  # "tools" or "tools_^_…"
+        output_pin_name.startswith("tools_^_") and "_~_" in output_pin_name
```

Comment:
use is_tool_pin ?

Reply:
`is_tool_pin` is less specific than this check